server/ot_rpc.go

```diff
@@ -0,0 +1,236 @@
+
+var SecretErr = errors.New("Invalid secret")
+
+// Transcoder
+
+func RunTranscoder(n *core.LivepeerNode, orchAddr string) {
+	expb := backoff.NewExponentialBackOff()
+	expb.MaxInterval = time.Minute
+	expb.MaxElapsedTime = 0
+	backoff.Retry(func() error {
+		glog.Info("Registering transcoder to ", orchAddr)
+		err := runTranscoder(n, orchAddr)
+		glog.Info("Unregistering transcoder: ", err)
+		if err != SecretErr {
+			return err
+		}
+		glog.Info("Terminating transcoder")
+		return nil
+	}, expb)
+}
+
+func runTranscoder(n *core.LivepeerNode, orchAddr string) error {
+	tlsConfig := &tls.Config{InsecureSkipVerify: true}
+	conn, err := grpc.Dial(orchAddr,
+		grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)))
+	if err != nil {
+		glog.Error("Did not connect transcoder to orchesrator: ", err)
+		return err
+	}
+	defer conn.Close()
+
+	c := net.NewTranscoderClient(conn)
+	ctx := context.Background()
+	r, err := c.RegisterTranscoder(ctx, &net.RegisterRequest{Secret: n.OrchSecret})
+	if err != nil {
+		glog.Error("Could not register transcoder to orchestrator ", err)
+		status := status.Convert(err)
+		if status.Message() == SecretErr.Error() { // consider this unrecoverable
+			return SecretErr
+		}
+		return err
+	}
+
+	httpc := &http.Client{Transport: &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}}
+	for {
+		notify, err := r.Recv()
+		if err != nil {
+			status := status.Convert(err)
+			if status.Message() == SecretErr.Error() { // consider this unrecoverable
+				return SecretErr
+			}
+			return err
+		}
+		profiles, err := common.TxDataToVideoProfile(hex.EncodeToString(notify.Profiles))
+		if err != nil {
+			glog.Info("Unable to deserialize profiles ", err)
+		}
+
+		glog.Info("Transcoding ", notify.TaskId)
+		var contentType string
+		var body bytes.Buffer
+
+		tData, err := n.Transcoder.Transcode(notify.Url, profiles)
+		if err != nil {
+			glog.Error("Unable to transcode ", err)
+			body.Write([]byte(err.Error()))
+			contentType = TranscodingErrorMimeType
+		} else {
+			boundary := randName()
+			w := multipart.NewWriter(&body)
+			for _, v := range tData {
+				w.SetBoundary(boundary)
+				hdrs := textproto.MIMEHeader{
+					"Content-Type":   {"video/MP2T"},
+					"Content-Length": {strconv.Itoa(len(v))},
+				}
+				fw, err := w.CreatePart(hdrs)
+				if err != nil {
+					glog.Error("Could not create multipart part ", err)
+					return err // XXX respond w error to orchestrator
+				}
+				io.Copy(fw, bytes.NewBuffer(v))
+			}
+			w.Close()
+			contentType = "multipart/mixed; boundary=" + boundary
+		}
+		req, err := http.NewRequest("POST", "https://"+orchAddr+"/transcodeResults", &body)
+		if err != nil {
+			glog.Error("Error posting results ", err)
+			return err
+		}
+		req.Header.Set("Authorization", ProtoVer_LPT)
+		req.Header.Set("Credentials", n.OrchSecret)
+		req.Header.Set("Content-Type", contentType)
+		req.Header.Set("TaskId", strconv.FormatInt(notify.TaskId, 10))
+		httpc.Do(req)
```

Comment:
Shouldn't we close response's body here?

Reply:
Ah yeah, I always forget that responses need to be cleaned up manually. Fixed.